runtime/src/iree/hal/driver.h

```diff
@@ -70,13 +70,62 @@ IREE_API_EXPORT iree_status_t iree_hal_driver_query_available_devices(
-IREE_API_EXPORT iree_status_t iree_hal_driver_create_device(
+// Creates a device with the given |device_ordinal| as enumerated by
+// iree_hal_driver_query_available_devices. The ordering of devices is driver
+// dependent and may vary across driver instances and processes.
+//
+// Optionally takes a list of key-value parameters as defined by the device
+// implementation. The parameters are only used during creation and need not
+// remain live after the function returns.
+IREE_API_EXPORT iree_status_t iree_hal_driver_create_device_by_ordinal(
+    iree_hal_driver_t* driver, iree_host_size_t device_ordinal,
+    iree_host_size_t param_count, const iree_string_pair_t* params,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device);
+
+// Creates a device with the given |device_id| obtained from
+// iree_hal_driver_query_available_devices. The |device_id| is an opaque value
+// that is stable for the lifetime of the driver but not across driver instances
+// or across processes. Only IDs as queried from the driver may be used. If an
+// ID of `IREE_HAL_DEVICE_ID_DEFAULT` is provided the driver will select a
+// device based on heuristics, flags, the environment, or a die roll.
+//
+// Optionally takes a list of key-value parameters as defined by the device
+// implementation. The parameters are only used during creation and need not
+// remain live after the function returns.
+IREE_API_EXPORT iree_status_t iree_hal_driver_create_device_by_id(
     iree_hal_driver_t* driver, iree_hal_device_id_t device_id,
+    iree_host_size_t param_count, const iree_string_pair_t* params,
+    iree_allocator_t host_allocator, iree_hal_device_t** out_device);
+
+// Creates a device with the given device path specifier.
+// The |device_path| is a driver-dependent string that identifies a particular
+// device or type of device. If the path is omitted the driver will select a
+// device based on heuristics, flags, the environment, or a die roll.
+//
+// Optionally takes a list of key-value parameters as defined by the device
+// implementation. The parameters are only used during creation and need not
+// remain live after the function returns.
+IREE_API_EXPORT iree_status_t iree_hal_driver_create_device_by_path(
+    iree_hal_driver_t* driver, iree_string_view_t driver_name,
+    iree_string_view_t device_path, iree_host_size_t param_count,
+    const iree_string_pair_t* params, iree_allocator_t host_allocator,
+    iree_hal_device_t** out_device);
+
+// Creates a device with the given `driver://path?params` URI.
+// The driver specifier is required in the URI but the path and params are
+// optional. The path is a driver-dependent string that identifies a particular
+// device or type of device. If the path is omitted the driver will select a
+// device based on heuristics, flags, the environment, or a die roll.
+//
+// This is equivalent to performing URI parsing and calling
+// iree_hal_driver_create_device_by_path.
```

Comment:
Once a few drivers use the paths and parameters, where will the driver-dependent strings be documented? A few examples (for driver implementers to reference) in here would also be nice IMO.

Reply:
They'll be documented with the drivers.
I'm the one who's going to be adding them first and I'll add examples when I figure out what I want :)